reactioncommerce/reaction-next-starterkit
# Simplify customization

Resolves #438 
Resolves #409 
Impact: **breaking**
Type: **feature**

Various changes based on audits and feedback about difficulties running, building, and customizing.

## Changes
- Moved `.yarnrc` to `.reaction/yarnrc-docker.template` and updated the Dockerfile to copy it in. This matches how it's done in component-library and fixes the fact that the yarn config prevents yarn from working outside of Docker.
- Other changes to Dockerfile to match the one in component-library
- Move analytics from `/config` to `/src/custom`
- Move `next.config.js` into `/src`.
- Fix NextJS build config so that all of the manual steps are no longer needed.
- Run the dev server with normal `node` rather than `babel-node`. `babel-node` was unnecessary and slower.
- Update to latest `@reactioncommerce/components-context`
- Add `jest-styled-components` for tests. This updated all of the snapshots so that styles are also snapshotted.
- Move `setupTests.js` from `/src` to `/tests/`
- Add `envalid` and update everything that uses `process.env` so that is uses config that is run through envalid for validation and transformation first.
- Remove remaining Keycloak code that was never used
- Remove some properties of the AuthStore that are no longer used
- Move `reactionTheme.js` and `componentTheme.js` to `/src/custom`
- Move the components context file from `/src/lib/theme/components.js` to `/src/custom/componentsContext.js`
- Move route definition to a new `defineRoutes` function in `/src/custom/routes.js`
- Clean up the default `logger` object a bit
- Remove unused `appConfig` from MobX `UIStore`
- Move `pagination` functions from `helpers` to `utils` and get rid of `helpers` folder
- Remove unused props from `serverRuntimeConfig` and `publicRuntimeConfig`
- Fix errors on the `_error` page
- Move all auth-related server setup out of `server.js` to a new `serverAuth.js`
- Server now listens on `PORT` env variable, which defaults to 4000 if not set, rather than being hard-coded to 4000.
- Remove the `request` NPM package and use `fetch` instead in the two places where `request` was used. This removes a large dependency that was unnecessary since `fetch` is already available.
- Fix #409 

## Breaking changes
Because some of the build steps and file locations, and the way env variables are handled, have changed, this may break custom scripts or tools. In general the defaults here should be much simpler now.

**After pulling this in for local development, remove all volumes that begin with `reaction-next-starterkit` (`docker volume rm reaction-next-starterkit_empty_node_modules reaction-next-starterkit_node_modules reaction-next-starterkit_web-yarn`) and rebuild the image the first time you run the project (`docker-compose up -d --build`). Also verify that your `.env` file has all of the necessary variables based on `.env.example`**

If you have server code that was relying on the fact that `babel-node` was transpiling it, you'll need to update it to syntax that is compatible with Node 10.

## Testing
- Verify all major aspects of the app continue to work correctly when running in dev mode (`docker-compose up -d --build`)
- Verify that you can build for production (see instructions in readme)
- Verify all major aspects of the app continue to work correctly when running the production Docker image you built.
- In particular, test logging in and out.
- Run without env variables set and verify that you get helpful errors on startup now.
